cloud/amazon/ec2_vpc_dhcp_options.py

```diff
@@ -128,105 +157,204 @@
     inherit_existing: True
     delete_old: False
+
+
+## Create a DHCP option set with 4.4.4.4 and 8.8.8.8 as the specified DNS servers, with tags
+## but do not assign to a VPC
+- ec2_vpc_dhcp_options:
+    region: us-east-1
+    dns_servers:
+      - 4.4.4.4
+      - 8.8.8.8
+    tags:
+      Name: google servers
+      Environment: Test
+
+## Delete a DHCP options set that matches the tags and options specified
+- ec2_vpc_dhcp_options:
+    region: us-east-1
+    dns_servers:
+      - 4.4.4.4
+      - 8.8.8.8
+    tags:
+      Name: google servers
+      Environment: Test
+  state: absent
+
+## Associate a DHCP options set with a VPC by ID
+- ec2_vpc_dhcp_options:
+    region: us-east-1
+    dhcp_options_id: dopt-12345678
+    vpc_id: vpc-123456
+
 """
 
 import boto.vpc
+import boto.ec2
+from boto.exception import EC2ResponseError
 import socket
 import collections
 
-def _get_associated_dhcp_options(vpc_id, vpc_connection):
+def get_resource_tags(vpc_conn, resource_id):
+    return dict((t.name, t.value) for t in vpc_conn.get_all_tags(filters={'resource-id': resource_id}))
+
+def ensure_tags(vpc_conn, resource_id, tags, add_only, check_mode):
+    try:
+        cur_tags = get_resource_tags(vpc_conn, resource_id)
+        if tags == cur_tags:
+            return {'changed': False, 'tags': cur_tags}
+
+        to_delete = dict((k, cur_tags[k]) for k in cur_tags if k not in tags)
+        if to_delete and not add_only:
+            vpc_conn.delete_tags(resource_id, to_delete, dry_run=check_mode)
+
+        to_add = dict((k, tags[k]) for k in tags if k not in cur_tags)
+        if to_add:
+            vpc_conn.create_tags(resource_id, to_add, dry_run=check_mode)
+
+        latest_tags = get_resource_tags(vpc_conn, resource_id)
+        return {'changed': True, 'tags': latest_tags}
+    except EC2ResponseError as e:
+        raise Exception('Unable to update tags for {0}, error: {1}'.format(resource_id, e))
+
+def fetch_dhcp_options_for_vpc(vpc_conn, vpc_id):
     """
     Returns the DHCP options object currently associated with the requested VPC ID using the VPC
     connection variable.
     """
-    vpcs = vpc_connection.get_all_vpcs(vpc_ids=[vpc_id])
+    vpcs = vpc_conn.get_all_vpcs(vpc_ids=[vpc_id])
     if len(vpcs) != 1:
-      return None
-    dhcp_options = vpc_connection.get_all_dhcp_options(dhcp_options_ids=[vpcs[0].dhcp_options_id])
+        return None
+    dhcp_options = vpc_conn.get_all_dhcp_options(dhcp_options_ids=[vpcs[0].dhcp_options_id])
     if len(dhcp_options) != 1:
-      return None
+        return None
     return dhcp_options[0]
 
+def match_dhcp_options(vpc_conn, tags=None, options=None):
+    """
+    Finds a DHCP Options object that optionally matches the tags and options provided
+    """
+    dhcp_options = vpc_conn.get_all_dhcp_options()
+    for dopts in dhcp_options:
+        if tags and get_resource_tags(vpc_conn, dopts.id) == tags:
+            if options and dopts.options == options:
```

Comment:
If no tags are provided, then this matches only on options, if no options were provided it matches only on tags

imo, from a 'dwim' perspective, matching the first found is the right behaviour - the request was for a dhcp_options that matches, and one was found... if they're looking to hit a specific one they could specify the id.

(I can see it being tricky in the case of stat=absent, rather than creation/assignment - but since dhcp options can't be deleted if they are assigned to a vpc, it isn't dangerous per se... worst case is that something *doesn't* get deleted)